TypeFunction.callMatch: Improve structure and control flow

Geod24 opened the pull request:
Just a refactoring, but a needed one.

Geod24 committed:
TypeFunction: Turn else { if / else } into 'else if' + 'else'

Reduces indentation and make the code more streamlined.

thewilsonator commented on compiler/src/dmd/mtype.d:
```diff
@@ -7343,3 +7192,157 @@ private extern(D) bool isCopyConstructorCallable (StructDeclaration argStruct,
+    }
+
+    // Non-lvalues do not match ref or out parameters
+    if (p.isReference())
```
could use an early return here.

Geod24 replied:
I don't think an early return makes the code clearer here. If anything, the reference code could be separated in its own function, which IMO would be better. But keep in mind I'm doing those refactors to make it easier to implement named parameters.

thewilsonator reviewed (approved):
Otherwise looks great!

Geod24 committed:
Trivial: Nest more variables inside switch/case in callMatch

Geod24 committed:
callMatch: Structure usage of trailing args in typesafe varargs

Geod24 committed:
callMatch: Remove goto that simply returns

Since those 'goto' where inside a 'if' that checked that
parameterList.varArgs was 2 (typesafe), it was impossible
that the 'if' would be entered, meaning the goto was
always returning, which is much easier to read.

Geod24 commented:
Reduced the scope of the PR to some smaller improvements, while I'm trying to untangle the rest.

thewilsonator commented:
I'll leave merging to you.
